include/mqtt/endpoint.hpp

```diff
@@ -8278,1217 +8325,3808 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
+            process_connack(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
+            return;
         case control_packet_type::publish:
             if (mqtt_connected_) {
-                ret = handle_publish(func);
+                process_publish(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::puback:
             if (mqtt_connected_) {
-                ret = handle_puback(func);
+                process_puback(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::pubrec:
             if (mqtt_connected_) {
-                ret = handle_pubrec(func);
+                process_pubrec(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::pubrel:
             if (mqtt_connected_) {
-                ret = handle_pubrel(func);
+                process_pubrel(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::pubcomp:
             if (mqtt_connected_) {
-                ret = handle_pubcomp(func);
+                process_pubcomp(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::subscribe:
             if (mqtt_connected_) {
-                ret = handle_subscribe(func);
+                process_subscribe(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::suback:
             if (mqtt_connected_) {
-                ret = handle_suback(func);
+                process_suback(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::unsubscribe:
             if (mqtt_connected_) {
-                ret = handle_unsubscribe(func);
+                process_unsubscribe(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::unsuback:
             if (mqtt_connected_) {
-                ret = handle_unsuback(func);
+                process_unsuback(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::pingreq:
             if (mqtt_connected_) {
-                ret = handle_pingreq(func);
+                process_pingreq(std::move(func));
             }
             break;
         case control_packet_type::pingresp:
             if (mqtt_connected_) {
-                ret = handle_pingresp(func);
+                process_pingresp(std::move(func));
```

Comment:
Hmm, I want to keep the current style for the parameters.
My policy is:

1. parameter is always used in the function, then the function has the  parameter.
1. parameter is always unnecessary,  then the function doesn't have the parameter.
1. parameter is sometimes but not always used,  then the function has the parameter.

process_pingreq and process_pingresp are the case 2, so I don't want to add the parameters.

Reply:
The reason why I made the suggestion, is that I was wondering if it made sense to have a small switch case, that decided what the type of the packet was, and set a pointer-to-member-function pointer to the next function to call.

Then, a single chunk of code can handle the logic for creating the buffer and everything involved in calling the next function.

I'm not sure that's 100% possible, but it wouldn't impact performance much (or at all), and it might improve readability.

But it would also require the function that gets called to have the same signature.

Reply:
Ok, I think that it is for additional performance improvement possibility. 
I'd like to keep current implementation on the PR. After it is merged, then we will try it.

I think that performance improvement should be checked using small test code before modify mqtt_cpp. Speaking of readability (and debuggability), it depends on situation. I usually set break point to this switch-case, it is sometimes helpful to understand what is happening.

Anyway, could you create an issue for that?

Reply:
My suggestion is not intended to be a performance improvement.

It's intended to enhance readability, and I suspect the performance impact will be minimal or not at all.

Reply:
I think that the original code is the same switch-case. I think that I don't need to fix it at this time.

Reply:
Ok. Maybe later.